BVAL-496 Improved java.time support

gsmet opened the pull request:
So this is the first part of a series of PRs related to java.time on BVAL, HV and BVTCK.

See beanvalidation/beanvalidation-tck#77 for the TCK changes.

See https://github.com/hibernate/hibernate-validator/pull/576 for the HV changes.

I hope it will be a good base for our future discussions.

This is based on the last version of https://github.com/sjmisterm/beanvalidation.org/blob/patch-1/proposals/BVAL-496.adoc .

A couple of differences though in the approach:
- I didn't use introspection to build the `now` object from the Clock and get the `compareTo` method. It would save some code but I don't think it's a good idea. I agree that it limits the support to officially supported types but I like it better this way. `AbstractJavaTimeValidator` makes it easy to add support for other types if required;
- I implemented all the default java.time types in this first round but I'll add optional support for threeten-extras as a proprietary extension in a followup-up patch if we decide to keep this approach;
- I named the `nowIsValid` option `orPresent` leading to having something like `@Past(orPresent = true)` which does not look too bad.

As for discussing it:
- remarks on the general approach should be discussed here so that they are centralized;
- comments on a specific PR should be added to the PR.

chkal commented:
Great work! Look VERY good. I like the idea of using `orPresent`. :+1:

hendrikebbers commented:
👍

gunnarmorling commented on src/main/java/javax/validation/ClockProvider.java:
```diff
@@ -0,0 +1,42 @@
+/*
+* Copyright 2009-2012, Red Hat, Inc. and/or its affiliates, and individual contributors
```
2016. In parallel need to check with @emmanuelbernard whether we can use the simplified form as in the RI, which avoids this issue to begin with.

gunnarmorling replied:
I've filed [BVAL-547](https://hibernate.atlassian.net/browse/BVAL-547) for the headers. We can use the simpler ones.

gunnarmorling commented on src/main/java/javax/validation/Configuration.java:
```diff
@@ -121,6 +122,20 @@
 	T parameterNameProvider(ParameterNameProvider parameterNameProvider);
 
 	/**
+	 * Defines the clock provider. Has priority over the configuration
```
What is the "configuration based" provider?

gsmet replied:
See the javadoc for `traversableResolver`. I supposed it refers to the one from the validation.xml conf file.

gunnarmorling replied:
Ok. Yes, I'd take it as that as well.

gunnarmorling commented on src/main/java/javax/validation/Configuration.java:
```diff
@@ -234,6 +249,20 @@
+	 * interface following the default {@code ClockProvider}
+	 * defined in the specification:
+	 *
+	 * XXX BVAL-496 paste definition of the specification
```
What is XXX? Maybe just omit this?

gsmet replied:
XXX is one of the convention for TODO that needs fixing before release. I used that consistently in the API/TCK/RI to mark all the places we need to update once the spec is written.

gunnarmorling replied:
Do you mind if i change that into "TODO". Seems a bit more common. Also this one is the only occurence I could find.

gsmet replied:
About TODO vs XXX, the idea is that XXX is something that really needs to be fixed (in my mind, it needs to be fixed before the release and nothing should be released with an XXX in it). A TODO might have a longer lifecycle and is less imperative. They are both supported by Eclipse.

gunnarmorling commented on src/main/java/javax/validation/ClockProvider.java:
```diff
@@ -0,0 +1,42 @@
+	/**
+	 * Returns the clock which serves as the reference for {@code now}.
+	 *
+	 * @return the clock which serves as the reference for {@code now}; Must not be {@code null}
```
"Must"="must"

gunnarmorling commented on src/main/java/javax/validation/constraints/Past.java:
```diff
@@ -64,6 +78,11 @@
 	/**
+	 * @return true if the present date should be considered valid
+	 */
+	boolean orPresent() default false;
```
What is the "granularity" of present? E.g. when validating

    @Past(orPresent=true) Year year = Year.of(2016);

Would that be true for every day in 2016?

gsmet replied:
Yes.

We build a corresponding `java.time` object from the `Clock` so to validate a `Year`, we build a `Year` object with the `Clock`. Thus 2016 would be valid for every day in 2016.

gsmet commented:
@gunnarmorling  force pushed a new version rebased and addressing most of your comments. About the `XXX`s, I would like to let them for now so that we don't forget to update the references to the spec. Either that or we decide we don't need to quote the spec which would be fine with me. I followed the existing convention.

gunnarmorling commented:
Merged, thanks to @gsmet for doing the change and to @hendrikebbers and @chkal for reviewing! I'll prepare the corresponding change to the spec now.

MenoData commented:
Has the generic support for any implementations of `TemporalAccessor` been dropped now? As far as I have understood, 3 requirements for an implementation class T were written down in the proposal:

- T implements TemporalAccessor
- T implements Comparable<T>
- has static `now(java.time.Clock)`-factory method yielding an instance of T

My own time library Time4J could easily offer more classes which meet these requirements (with a very few changes). And other library providers could also take profit (like Threeten-Extra mentioned in the proposal).

gunnarmorling commented:
@MenoData we went for a specific set of supported types, based on those requirements above.

The reason for not baking these requirements into the spec verbatim is that it'd require a reflective `ConstraintValidator` implementation which is something that shouldn't be mandated by the spec. Worse, though, the requirement `T exends TemporalAccessor & Comparable<T>` cannot be translated into a supported `ConstraintValidator` definition.

What you can do though is to provide constraint validators for your Time4J types as part of your library (declare the dependency to the BV API as optional). Then, when suing the BV reference implementation, the validators will be picked up via the service loader mechanism, provided you add a file `META-INF/services/javax.validation.ConstraintValidator` with the FQNs of your implementations to your JAR. We may add support for this to the spec itself, but I'm not sure whether we'll get to it in the 2.0 time frame.
